nemo/utils/exp_manager.py

```diff
@@ -307,11 +307,22 @@ def exp_manager(trainer: 'pytorch_lightning.Trainer', cfg: Optional[Union[DictCo
             exp_dir, name, and version.
     """
     # Add rank information to logger
-    # Note: trainer.global_rank and trainer.is_global_zero are not set until trainer.fit, so have to hack around it
     local_rank = int(os.environ.get("LOCAL_RANK", 0))
     global_rank = trainer.node_rank * trainer.num_devices + local_rank
     logging.rank = global_rank
 
+    # the exp manager calls some operations that require explicit
+    # synchronization, therefore we need to initialize the process
+    # group to initiate a barrier
+    if not torch.distributed.is_initialized() and trainer.num_nodes * trainer.num_devices > 1:
+
+        def dummy():
```

Comment:
Hmm that seems reasonable to avoid race conditions.

It also should probably not have an effect on 1 GPU things. 

However I'm worried about what hacking ptl like this will do to ddp training and stuff. What happens when you call trainer.fit() after already setting up ddp with that dummy ? Will it reinitialize ? Not likely. So what does the dummy do then to normal PTL behaviour ?

Reply:
when testing on my local setup, doing trainer.fit after this initialization seems totally fine. But I don't really know how it behaves on larger runs. To be save will you rather me modify NLPDDPStrategy as well and set a flag that checks if setup_environment has already been called?

Reply:
actually looking closely, i think this is safe because PTL guards against already initialized case  https://github.com/Lightning-AI/lightning/blob/984f49f7195ddc67e961c7c498ee6e19fc0cecb5/src/lightning/fabric/utilities/distributed.py#L237